core/amber/src/main/python/core/runnables/data_processor.py

```diff
@@ -29,40 +30,53 @@ def run(self) -> None:
         self._running.set()
         self._switch_context()
         while self._running.is_set():
-            self.process_tuple()
+            marker = self._context.marker_processing_manager.get_input_marker()
+            tuple_ = self._context.tuple_processing_manager.current_input_tuple
+            if marker is not None:
+                self.process_marker(marker)
+            elif tuple_ is not None:
+                self.process_tuple()
+            else:
+                raise RuntimeError("No marker or tuple to process.")
+            self._switch_context()
+
+    def process_marker(self, marker: Marker) -> None:
+        try:
```

Comment:
as dp is the core module of the python engine, please make sure you document well. please add doc-strings to all the methods.